subprojects/gstreamer/plugins/tracers/gstframerate.c

```diff
@@ -0,0 +1,362 @@
+{
+  GstElement *peer_element;
+  guint64 min;
+  guint64 max;
+};
+
+/* data helpers */
+
+/*
+ * Get the element/bin owning the pad.
+ *
+ * in: a normal pad
+ * out: the element
+ *
+ * in: a proxy pad
+ * out: the element that contains the peer of the proxy
+ *
+ * in: a ghost pad
+ * out: the bin owning the ghostpad
+ */
+/* TODO(ensonic): gst_pad_get_parent_element() would not work here, should we
+ * add this as new api, e.g. gst_pad_find_parent_element();
+ */
+static GstElement *
+get_real_pad_parent (GstPad *pad)
+{
+  GstObject *parent;
+
+  if (!pad)
+    return NULL;
+
+  parent = gst_object_get_parent (GST_OBJECT_CAST (pad));
+
+  /* if parent of pad is a ghost-pad, then pad is a proxy_pad */
+  if (parent && GST_IS_GHOST_PAD (parent)) {
+    GstObject *tmp;
+    pad = GST_PAD_CAST (parent);
+    tmp = gst_object_get_parent (GST_OBJECT_CAST (pad));
+    gst_object_unref (parent);
+    parent = tmp;
+  }
+  return GST_ELEMENT_CAST (parent);
+}
+
+/* hooks */
+
+
+static gboolean
+gst_framerate_tracer_display_current_fps (GstFramerateTracer *self,
+    GstClockTime ts)
+{
+  guint64 frames_rendered, frames_dropped;
+  gdouble rr, dr, average_fps;
+  gchar fps_message[256];
+  gdouble time_diff, time_elapsed;
+  GstClockTime current_ts = gst_util_get_timestamp ();
+
+  frames_rendered = g_atomic_int_get (&self->frames_rendered);
+  frames_dropped = g_atomic_int_get (&self->frames_dropped);
+
+  if ((frames_rendered + frames_dropped) == 0) {
+    /* in case timer fired and we didn't yet get any QOS events */
+    return TRUE;
+  }
+
+  time_diff = (gdouble) (current_ts - self->last_ts) / GST_SECOND;
+  time_elapsed = (gdouble) (current_ts - self->start_ts) / GST_SECOND;
+
+  rr = (gdouble) (frames_rendered - self->last_frames_rendered) / time_diff;
+  dr = (gdouble) (frames_dropped - self->last_frames_dropped) / time_diff;
+
+  average_fps = (gdouble) frames_rendered / time_elapsed;
+
+  if (self->max_fps == -1 || rr > self->max_fps) {
+    self->max_fps = rr;
+    GST_DEBUG_OBJECT (self, "Updated max-fps to %f", rr);
+  }
+  if (self->min_fps == -1 || rr < self->min_fps) {
+    self->min_fps = rr;
+    GST_DEBUG_OBJECT (self, "Updated min-fps to %f", rr);
+  }
+
+  /* Display on a single line to make it easier to read and import
+   * into, for example, excel..  note: it would be nice to show
+   * timestamp too.. need to check if there is a sane way to log
+   * timestamp of last rendered buffer, so we could correlate dips
+   * in framerate to certain positions in the stream.
+   */
+  if (dr == 0.0) {
+    g_snprintf (fps_message, 255,
+        "rendered: %" G_GUINT64_FORMAT ", dropped: %" G_GUINT64_FORMAT
+        ", current: %.2f, average: %.2f", frames_rendered, frames_dropped, rr,
+        average_fps);
+  } else {
+    g_snprintf (fps_message, 255,
+        "rendered: %" G_GUINT64_FORMAT ", dropped: %" G_GUINT64_FORMAT
+        ", fps: %.2f, drop rate: %.2f", frames_rendered, frames_dropped, rr,
+        dr);
```

Comment:
what's the point of these lines? We fill `fps_message`, but not use it anywhere.

P.S. [YAGNI](https://es.wikipedia.org/wiki/YAGNI)